project-koku/koku
# 1912 tags endpoint

#1912

Adds the `tags/{infra}/{key}/` endpoint.
Since I was already modifying the urls.py file, I switched out `url()` for `path()` because `url()` may be deprecated in the future ([link](https://docs.djangoproject.com/en/2.2/ref/urls/#url)).

Testing:
1. load some tagged test data.
2. Go to `tags/{infra}/` and take one of the `key`s.
3. Take the key, and go to `tags/{infra}/{key}/.
4. See a populated list of values associated with that key.
5. Apply random filters/group_bys/order_bys. See that the all work.
6. Apply key_only=True, see Bad Request.
7. Try to hit an endpoint for a tag that does not exist, see 404 not found.

results:
```
GET /api/cost-management/v1/tags/aws/

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/?limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/?limit=100&offset=0"
    },
    "data": [
        {
            "key": "version",
            "values": [
                "prod",
                "master",
                "gamma",
                "beta"
            ]
        },
...
```

```
GET /api/cost-management/v1/tags/aws/version/

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/version/?limit=100&offset=0"
    },
    "data": [
        "prod",
        "master",
        "gamma",
        "beta"
    ]
}
```
If using the default Nise data, apply: `?filter[account]=8`
```
GET /api/cost-management/v1/tags/aws/version/?filter[account]=8

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 0,
        "filter": {
            "account": [
                "8"
            ],
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "key_only": false,
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?filter%5Baccount%5D=8&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/version/?filter%5Baccount%5D=8&limit=100&offset=0"
    },
    "data": []
}
```
Apply non-sense:
``` 
GET /api/cost-management/v1/tags/aws/non-sense-tag-which-does-not-exist/

HTTP 404 Not Found
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "errors": [
        {
            "detail": "Not found.",
            "source": "detail",
            "status": 404
        }
    ]
}
```
Try to apply the `key_only` tag query param:
```
GET /api/cost-management/v1/tags/aws/version/?key_only=True
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "errors": [
        {
            "detail": "Invalid query parameter 'key_only'.",
            "source": "details",
            "status": 400
        }
    ]
}
```
